Add hipclang amdgcn functions

aaronenyeshi opened the pull request:
These are moving from hipclang in device library to hip headers. These are required for the functionality of HIPclang project.

aaronenyeshi commented:
So far I have added assertfail, ballot, clock, lanemask, memory, named sync, smid, sync, and trap. These previously existed as .cl and .ll files in https://github.com/RadeonOpenCompute/ROCm-Device-Libs/tree/9b46f23b1795b56600bb39ff2594ce129ab0222e/hip/src

yxsamliu commented on include/hip/hcc_detail/hip_runtime.h:
```diff
@@ -656,8 +713,136 @@ __DEVICE__ void inline __assert_fail(const char * __assertion,
+
+// hip.amdgcn.bc - sync threads
+// extern "C" __device__ __attribute__((noduplicate)) void __syncthreads();
+#define CLK_LOCAL_MEM_FENCE    0x01
```
please prefix with __, since we don't want to pollute users' name space

yxsamliu commented on include/hip/hcc_detail/hip_runtime.h:
```diff
@@ -656,8 +713,136 @@ __DEVICE__ void inline __assert_fail(const char * __assertion,
+// hip.amdgcn.bc - sync threads
+// extern "C" __device__ __attribute__((noduplicate)) void __syncthreads();
+#define CLK_LOCAL_MEM_FENCE    0x01
+#define local __attribute__((address_space(3)))
```
prefix local with __

yxsamliu commented on include/hip/hcc_detail/hip_runtime.h:
```diff
@@ -656,8 +713,136 @@ __DEVICE__ void inline __assert_fail(const char * __assertion,
+#define CLK_LOCAL_MEM_FENCE    0x01
+#define local __attribute__((address_space(3)))
+
+typedef unsigned cl_mem_fence_flags;
```
prefix with __

yxsamliu commented on include/hip/hcc_detail/hip_runtime.h:
```diff
@@ -656,8 +713,136 @@ __DEVICE__ void inline __assert_fail(const char * __assertion,
+
+typedef unsigned cl_mem_fence_flags;
+
+typedef enum memory_scope {
```
prefix with __

yxsamliu commented on include/hip/hcc_detail/hip_runtime.h:
```diff
@@ -656,8 +713,136 @@ __DEVICE__ void inline __assert_fail(const char * __assertion,
+} memory_scope;
+
+// enum values aligned with what clang uses in EmitAtomicExpr()
+typedef enum memory_order
```
prefix with __

yxsamliu commented on include/hip/hcc_detail/hip_runtime.h:
```diff
@@ -313,8 +360,11 @@ __device__ int __hip_move_dpp(int src, int dpp_ctrl, int row_mask, int bank_mask
 
+__device__
+inline
+void* __get_dynamicgroupbaseptr() { return get_dynamic_group_segment_base_pointer(); }
```
can we remove get_dynamic_group_segment_base_pointer and implement __get_dynamicgroupbaseptr directly? I think the disabled code in __amdgcn_get_dynamicgroupbaseptr is related to this.

yxsamliu commented on include/hip/hcc_detail/hip_runtime.h:
```diff
@@ -656,8 +713,158 @@ __DEVICE__ void inline __assert_fail(const char * __assertion,
+
+__device__
+inline
+static void hc_work_group_barrier(__cl_mem_fence_flags flags, __memory_scope scope)
```
can we rename it to __work_group_barrier?

yxsamliu commented on include/hip/hcc_detail/hip_runtime.h:
```diff
@@ -656,8 +713,158 @@ __DEVICE__ void inline __assert_fail(const char * __assertion,
+
+__device__
+inline
+static void hc_barrier(int n)
```
can we rename it as __barrier?

yxsamliu commented:
I think it is cleaner to create two new header files:

llvm_intrinsics.h - containing declarations for wrapper functions for llvm intrinsics like llvm.amdgcn.s.barrier

device_library_decls.h - containing declarations for types and functions in device library

Other changes in this patch can be moved to device_functions.h

aaronenyeshi commented:
I'm working on separating this into 3 files, llvm_intrinsics.h, device_library_decls.h, and device_functions.h. Thanks @yxsamliu

aaronenyeshi commented:
It doesn't seem that

aaronenyeshi commented:
I've rebased commits on master, and removed device_functions.cpp and hip_hc.ll and added them into HIP headers. Also working on removing ISA asm.

aaronenyeshi commented:
I'm not sure why 2 tests are failing: directed_tests/deviceLib/hipTestHalf and directed_tests/deviceLib/hipTestNativeHalf

mangupta commented on include/hip/hcc_detail/device_library_decls.h:
```diff
@@ -0,0 +1,62 @@
+
+// Introduce local address space
+#define __local __attribute__((address_space(3)))
+__device__ inline static __local char* __to_local(unsigned x) { return (__local char*)x; }
```
HCC throws the warning:
```warning: cast to '__attribute__((address_space(3))) char *' from smaller integer type 'unsigned int' [-Wint-to-pointer-cast]```

aaronenyeshi replied:
Should I hide this warning by adding CPP flag -Wno-int-to-pointer-cast ?

scchan replied:
The signature of this function doesn't look right to me.  I would expect `x` to be a generic pointer which is 64-bit.  Is there a unit test for this

yxsamliu replied:
This is a workaround for a compiler bug. It relies on compiler inserting an addrspacecast when calling this function.

Aaron, since this relies on a compiler workaround which is not in amd-common branch, there is no point of putting this change into HIP. Can you separate this change (and those depending on it) and keep it in hip-clang branch only? Then you can update this pull request without this part of the change.

yxsamliu replied:
sorry I mistaken this function with another function.

This is not the OpenCL __to_local function. It is a static function which is called by  __get_dynamicgroupbaseptr.

yxsamliu replied:
The warning comes from host compile stage. Although there is no IR emitted for this device function in host compilation stage, it still needs to be parsed. Then a pointer in addrspace 3 is 64 bit for host, which causes the warning. Since this function is valid only for device compilation, we should guard it with __HIP__DEVICE_COMPILE.

aaronenyeshi replied:
I will guard the original __to_local with `__HIP_DEVICE_COMPILE__` since a local pointer should be the same size as uint32_t on device side.

aaronenyeshi replied:
Seems if I surround __to_local with that guard, I will need to surround get_dynamicgroupbaseptr caller functions as well.

aaronenyeshi commented:
@mangupta I've merged to master resolving conflicts. Also I've added a fix for the warning message by using larger integer type before addrspacecast. Passing on my end and Jenkins now.

@scchan Thanks, I've fixed the arg `x` to be a generic unsigned 64 bit int before the addrspacecast now.

@yxsamliu It was the `__local_to_generic` function that encounters clang bug with addrspacecast. `__to_local` works fine in C++ compiler.

scchan commented on include/hip/hcc_detail/device_library_decls.h:
```diff
@@ -45,7 +45,8 @@ extern "C" __device__ float __ocml_trunc_f32(float);
 #define __local __attribute__((address_space(3)))
-__device__ inline static __local char* __to_local(unsigned x) { return (__local char*)x; }
+// TODO: Remove cast to uint64_t after clang bug is fixed. Should use uint32_t
+__device__ inline static __local char* __to_local(uint64_t x) { return (__local char*)x; }
```
should the return type be `void*` instead of `char*`?

aaronenyeshi replied:
Thanks, I will change that

aaronenyeshi commented:
@mangupta I've added a few more changes to allow hip-clang path to pass tests: hipSurfaceObj2D, hip_test_ldg, hip_threadfence_system, and hipNullStream. Thanks!

yxsamliu commented on include/hip/hcc_detail/device_functions.h:
```diff
@@ -142,72 +516,466 @@ CUDA implements half as unsigned short whereas, HIP doesn't.
+    return ballot;
+}
+
+#ifdef __HIP_DEVICE_COMPILE__
```
we still need the device functions to be declared for host compilation, therefore the code should be like this
`#ifdef __HIP_DEVICE_COMPILE__
__device__
inline
void* __get_dynamicgroupbaseptr()
{
    // Get group segment base pointer.
    return (char*)__local_to_generic(__to_local(__llvm_amdgcn_groupstaticsize()));
}
#else
__device__
void* __get_dynamicgroupbaseptr();
#endif // __HIP_DEVICE_COMPILE__
`

aaronenyeshi replied:
Hi Sam, I've cherry-picked your fixes into this PR related to your review. Thanks!

aaronenyeshi commented:
@yxsamliu @scchan @mangupta I believe this is ready to land. Please help review. It passes all unit tests on HCC/HIP, also it passes 97/106 unit tests on HIP-clang on HCC rt. Thank you!

yxsamliu commented:
Pleas hold on. We need Brian to review it. Thanks.

aaronenyeshi commented:
@b-sumner - I will make a new commit to remove the old usage of __activelanemake_v4_b64_b1.

b-sumner commented on include/hip/hcc_detail/device_library_decls.h:
```diff
@@ -0,0 +1,78 @@
+
+extern "C" __device__ int32_t __ockl_activelane_u32(void);
+
+extern "C" __device__ uint __ockl_mul24_u32(uint, uint);
```
Please copy the function attributes these functions are given in ocml.h.

b-sumner commented on include/hip/hcc_detail/hip_runtime.h:
```diff
@@ -235,85 +201,15 @@ __device__ int __hip_move_dpp(int src, int dpp_ctrl, int row_mask, int bank_mask
-//__device__ void  __threadfence_system(void) __attribute__((deprecated("Provided with workaround
-//configuration, see hip_kernel_language.md for details")));
-__device__ void __threadfence_system(void);
+__device__ inline static int min(int arg1, int arg2) {
```
These can't work.  An f32 can not exactly represent every i32.  Just use arg1 < arg2 ? arg1 : arg2.  The compiler will do the rest.

b-sumner commented on include/hip/hcc_detail/llvm_intrinsics.h:
```diff
@@ -0,0 +1,68 @@
+
+#include "hip/hcc_detail/host_defines.h"
+
+__device__ ulong __llvm_amdgcn_icmp_i32(uint x, uint y, uint z) __asm("llvm.amdgcn.icmp.i32");
```
This should have the convergent attribute if it is really needed.

aaronenyeshi commented on include/hip/hcc_detail/device_functions.h:
```diff
@@ -142,72 +516,457 @@ CUDA implements half as unsigned short whereas, HIP doesn't.
+__device__ static inline int __float2int_ru(float x) { return (int)__ocml_ceil_f32(x); }
+__device__ static inline int __float2int_rz(float x) { return (int)__ocml_trunc_f32(x); }
+
+__device__ static inline long long int __float2ll_rd(float x) { return (long long int)x; }
```
I copied these conversion functions from old HIP implementation. How can I fix it?

yxsamliu replied:
I think the correct versions cause isel failures in backend. Probably just add ToDo.

aaronenyeshi commented:
Thanks @b-sumner, I've added the missing attributes, changed llvm.amdgcn.icmp.i32 asm and changed min/max as you suggested. For the conversion functions, I've left a TODO comment for the future.

b-sumner commented:
Looks good to me.  I will work on the conversion functions after this is checked in.

aaronenyeshi commented:
@mangupta ping - we need this merged for internal HIP testing and investigation.
